File: src/fetch/scheduler.rs
use std::sync::Arc;

use chrono::Utc;
use tokio::time::{self, Duration};

use crate::config::GroupConfig;
use crate::fetch::fetcher;
use crate::generate;
use crate::mihomo_api;
use crate::model::GroupStatus;
use crate::port;
use crate::state::AppState;

/// Spawn background tasks that periodically refresh each subscription group.
pub fn spawn_refresh_tasks(state: AppState, client: reqwest::Client) {
    let config = Arc::clone(&state.config);
    for group_cfg in &config.groups {
        let state = state.clone();
        let client = client.clone();
        let group_cfg = group_cfg.clone();
        tokio::spawn(async move {
            refresh_loop(state, client, group_cfg).await;
        });
    }
}

async fn refresh_loop(state: AppState, client: reqwest::Client, group_cfg: GroupConfig) {
    let interval = Duration::from_secs(group_cfg.update_interval);
    let group_name = group_cfg.name.clone();

    // Initial fetch immediately
    do_refresh(&state, &client, &group_cfg).await;

    let mut ticker = time::interval(interval);
    ticker.tick().await; // skip the first immediate tick
    loop {
        ticker.tick().await;
        tracing::info!(group = %group_name, "scheduled refresh");
        do_refresh(&state, &client, &group_cfg).await;
    }
}

/// Perform a single refresh for one group: fetch, parse, assign ports, regenerate configs.
#[tracing::instrument(skip_all, fields(group = %group_cfg.name))]
pub async fn do_refresh(state: &AppState, client: &reqwest::Client, group_cfg: &GroupConfig) {
    match fetcher::fetch_group(client, group_cfg).await {
        Ok(mut nodes) => {
            let mut inner = state.inner.write().await;

            // Remove old entries for this group's previous nodes first
            if let Some(gs) = inner.groups.get(&group_cfg.name) {
                let old_names: Vec<String> = gs.nodes.iter().map(|n| n.name.clone()).collect();
                for name in old_names {
                    inner.port_map.remove(&name);
                }
            }

            // Assign ports (merge into global port_map)
            let new_entries = port::assign_ports(
                &mut nodes,
                state.config.port.range_start,
                &inner.port_map,
            );

            // Merge new entries
            inner.port_map.extend(new_entries);

            // Update group state
            if let Some(gs) = inner.groups.get_mut(&group_cfg.name) {
                gs.nodes = nodes;
                gs.last_updated = Some(Utc::now());
                gs.last_error = None;
                gs.status = GroupStatus::Ok;
            }

            tracing::info!("group updated successfully");

            // Regenerate configs (while still holding lock for consistency)
            let all_nodes: Vec<_> = inner
                .groups
                .values()
                .flat_map(|g| g.nodes.iter().cloned())
                .collect();
            drop(inner);

            if let Err(e) = generate::regenerate(&state.config, &all_nodes).await {
                tracing::error!(error = %e, "failed to regenerate configs");
            }

            // Reload mihomo
            if let Err(e) = mihomo_api::reload_config(client, &state.config.mihomo).await {
                tracing::warn!(error = %e, "failed to reload mihomo (may not be running)");
            }
        }
        Err(e) => {
            tracing::warn!(error = %e, "fetch failed, keeping cached data");
            let mut inner = state.inner.write().await;
            if let Some(gs) = inner.groups.get_mut(&group_cfg.name) {
                gs.last_error = Some(e.to_string());
                gs.status = GroupStatus::Error;
            }
        }
    }
}

File: zig/src/fetch/scheduler.zig
const std = @import("std");
const state = @import("../state.zig");
const config = @import("../config.zig");
const model = @import("../model.zig");
const fetcher = @import("fetcher.zig");
const generate = @import("../generate/mihomo.zig");
const port = @import("../port.zig");
const mihomo_api = @import("../mihomo_api.zig");

pub fn spawnRefreshTasks(app: *state.AppState, client: *std.http.Client) void {
    const gpa = app.gpa;
    const io = app.io;
    for (app.config.groups) |group| {
        const group_copy = tryCopyGroupConfig(gpa, group) catch continue;
        _ = io.async(refreshLoop, .{ app, client, group_copy });
    }
}

pub fn spawnSingleRefreshTask(app: *state.AppState, client: *std.http.Client, group: config.GroupConfig) void {
    const group_copy = tryCopyGroupConfig(app.gpa, group) catch return;
    _ = app.io.async(refreshLoop, .{ app, client, group_copy });
}

fn refreshLoop(app: *state.AppState, client: *std.http.Client, group: config.GroupConfig) void {
    const gpa = app.gpa;
    const io = app.io;

    // Initial refresh using the provided client
    doRefresh(app, client, &group) catch |err| { std.log.err("refresh failed for  ++ group.name ++ : {s}", .{@errorName(err)}); };

    while (true) {
        io.sleep(std.Io.Duration.fromSeconds(@intCast(group.update_interval)), .real) catch break;
        std.log.info("scheduled refresh: {s}", .{group.name});
        var fresh_client: std.http.Client = .{ .allocator = gpa, .io = io };
        defer fresh_client.deinit();
        doRefresh(app, &fresh_client, &group) catch |err| { std.log.err("refresh failed for  ++ group.name ++ : {s}", .{@errorName(err)}); };
    }

    // Cleanup copied config
    gpa.free(group.name);
    if (group.subscription) |s| gpa.free(s);
    if (group.file) |s| gpa.free(s);
    if (group.filter) |s| gpa.free(s);
    if (group.exclude_filter) |s| gpa.free(s);
}

pub fn doRefresh(app: *state.AppState, client: *std.http.Client, group: *const config.GroupConfig) !void {
    const gpa = app.gpa;
    const io = app.io;

    const nodes = fetcher.fetchGroup(gpa, io, client, group.*) catch |err| {
        std.log.warn("fetch failed for '{s}': {s}", .{ group.name, @errorName(err) });
        app.inner.lock();
        defer app.inner.unlock();
        if (app.inner_data.groups.getEntry(group.name)) |entry| {
            entry.value_ptr.last_error = std.fmt.allocPrint(gpa, "{s}", .{@errorName(err)}) catch null;
            entry.value_ptr.status = .error_status;
        }
        return;
    };

    app.inner.lock();

    // Remove old ports
    if (app.inner_data.groups.getEntry(group.name)) |entry| {
        for (entry.value_ptr.nodes) |*node| {
            _ = app.inner_data.port_map.remove(node.name);
        }
        entry.value_ptr.deinit(gpa);
    }

    // Assign ports
    const new_ports = port.assignPorts(gpa, nodes, app.config.port.range_start, &app.inner_data.port_map) catch |err| {
        std.log.err("port assignment failed: {s}", .{@errorName(err)});
        for (nodes) |*node| node.deinit(gpa);
        gpa.free(nodes);
        app.inner.unlock();
        return;
    };
    var pit = new_ports.iterator();
    while (pit.next()) |entry| {
        app.inner_data.port_map.put(gpa.dupe(u8, entry.key_ptr.*) catch {
            app.inner.unlock();
            return;
        }, entry.value_ptr.*) catch {
            app.inner.unlock();
            return;
        };
    }

    // Update group state
    const gs = model.GroupState{
        .name = gpa.dupe(u8, group.name) catch {
            app.inner.unlock();
            return;
        },
        .nodes = nodes,
        .last_updated = @intCast(@divTrunc(std.Io.Timestamp.now(io, .real).nanoseconds, std.time.ns_per_s)),
        .last_error = null,
        .status = .ok,
    };
    app.inner_data.groups.put(gpa, gpa.dupe(u8, group.name) catch {
        app.inner.unlock();
        return;
    }, gs) catch {
        app.inner.unlock();
        return;
    };

    // Regenerate configs
    const all_nodes = collectAllNodes(gpa, &app.inner_data) catch {
        app.inner.unlock();
        return;
    };
    defer gpa.free(all_nodes);

    app.inner.unlock();

    generate.generateMihomoConfig(gpa, io, app.config, all_nodes) catch |err| {
        std.log.err("failed to regenerate mihomo config: {s}", .{@errorName(err)});
    };
    mihomo_api.reloadConfig(gpa, io, client, app.config.mihomo) catch |err| {
        std.log.warn("failed to reload mihomo: {s}", .{@errorName(err)});
    };
}

fn collectAllNodes(gpa: std.mem.Allocator, inner: *state.AppStateInner) ![]model.ProxyNode {
    var total: usize = 0;
    var it = inner.groups.iterator();
    while (it.next()) |entry| {
        total += entry.value_ptr.nodes.len;
    }
    var all = try gpa.alloc(model.ProxyNode, total);
    var idx: usize = 0;
    it = inner.groups.iterator();
    while (it.next()) |entry| {
        @memcpy(all[idx..][0..entry.value_ptr.nodes.len], entry.value_ptr.nodes);
        idx += entry.value_ptr.nodes.len;
    }
    return all;
}

fn tryCopyGroupConfig(gpa: std.mem.Allocator, group: config.GroupConfig) !config.GroupConfig {
    return .{
        .name = try gpa.dupe(u8, group.name),
        .subscription = if (group.subscription) |s| try gpa.dupe(u8, s) else null,
        .file = if (group.file) |s| try gpa.dupe(u8, s) else null,
        .update_interval = group.update_interval,
        .filter = if (group.filter) |s| try gpa.dupe(u8, s) else null,
        .exclude_filter = if (group.exclude_filter) |s| try gpa.dupe(u8, s) else null,
    };
}

